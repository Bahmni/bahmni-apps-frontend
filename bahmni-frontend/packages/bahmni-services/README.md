# bahmni-services

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test bahmni-services` to execute the unit tests via [Vitest](https://vitest.dev/).
