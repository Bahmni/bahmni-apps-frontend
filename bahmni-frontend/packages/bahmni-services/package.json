{
  "name": "@bahmni-frontend/bahmni-services",
  "version": "0.0.1",
  "main": "./index.js",
  "types": "./index.d.ts",
  "exports": {
    ".": {
      "import": "./index.mjs",
      "require": "./index.js"
    }
  },
  "author": "Bahmni Team",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/bahnew/bahmni-clinical-frontend.git",
    "directory": "bahmni-frontend/packages/bahmni-services"
  },
  "dependencies": {
    "axios": "^1.6.0",
    "html-entities": "^2.4.0",
    "i18next": "^24.2.3",
    "i18next-browser-languagedetector": "^8.0.4"
  },
  "peerDependencies": {
    "react": ">=18.0.0",
    "react-dom": ">=18.0.0",
    "react-i18next": ">=13.0.0",
    "axios": ">=1.6.0 || < 2.0.0"
  }
}
