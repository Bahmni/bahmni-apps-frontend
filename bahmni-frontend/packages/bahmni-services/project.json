{
  "name": "bahmni-services",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/bahmni-services/src",
  "projectType": "library",
  "tags": ["type:lib", "scope:bahmni"],
  "targets": {
    "build": {
      "executor": "@nx/vite:build",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/bahmni-services"
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/bahmni-services/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": [
          "packages/bahmni-services/**/*.{ts,tsx,js,jsx}"
        ]
      }
    },
    "type-check": {
      "executor": "@nx/tsc:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "tsConfig": "packages/bahmni-services/tsconfig.lib.json",
        "outputPath": "dist/out-tsc/packages/bahmni-services"
      }
    }
  }
}
