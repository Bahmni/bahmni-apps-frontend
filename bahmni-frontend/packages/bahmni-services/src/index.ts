export { get, post, put, del } from './api';
export { initAppI18n, useTranslation } from './i18n';
export {
  getPatientById,
  getFormattedPatientById,
  type FormattedPatientData,
} from './patientService';
export { getFormattedError } from './errorHandling';
export { type Notification } from './notification';
