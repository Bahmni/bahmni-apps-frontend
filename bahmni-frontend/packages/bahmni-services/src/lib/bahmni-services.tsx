import styles from './bahmni-services.module.scss';

export function BahmniServices() {
  return (
    <div className={styles['container']}>
      <h1>Welcome to BahmniServices!</h1>
    </div>
  );
}

export default BahmniServices;
