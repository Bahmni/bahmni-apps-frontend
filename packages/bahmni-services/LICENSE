/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. Bahmni is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright 2025. Thoughtworks, Inc. Thoughtworks is a registered trademark 
 * and the Thoughtworks graphic logo is a trademark of Thoughtworks Inc.
 */
