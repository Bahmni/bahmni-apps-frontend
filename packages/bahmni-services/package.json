{
  "name": "@bahmni-frontend/bahmni-services",
  "version": "0.0.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "development": "./src/index.ts",
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "default": "./dist/index.js"
    }
  },
  "author": "Bahmni Team",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/bahnew/bahmni-clinical-frontend.git",
    "directory": "bahmni-frontend/packages/bahmni-services"
  },
  "dependencies": {
    "axios": "^1.12.0",
    "html-entities": "^2.6.0",
    "i18next": "^24.2.3",
    "react-i18next": "^15.4.1",
    "i18next-browser-languagedetector": "^8.0.4"
  },
  "peerDependencies": {
    "react": ">=18.0.0",
    "react-dom": ">=18.0.0"
  }
}
