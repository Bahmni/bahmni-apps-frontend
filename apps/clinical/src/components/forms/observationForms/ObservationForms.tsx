import {
  ComboBox,
  Tile,
  FormCard,
  FormCardContainer,
  SkeletonText,
} from '@bahmni-frontend/bahmni-design-system';
import { ObservationForm } from '@bahmni-frontend/bahmni-services';
import React, { useState, useMemo, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { DEFAULT_FORM_API_NAMES } from '../../../constants/forms';
import useObservationFormsSearch from '../../../hooks/useObservationFormsSearch';
import styles from './styles/ObservationForms.module.scss';

interface ObservationFormsProps {
  onFormSelect?: (form: ObservationForm) => void;
  selectedForms?: ObservationForm[];
  onRemoveForm?: (formUuid: string) => void;
  // Pinned forms state passed from parent (required)
  pinnedForms: ObservationForm[];
  updatePinnedForms: (newPinnedForms: ObservationForm[]) => Promise<void>;
  isPinnedFormsLoading: boolean;
}

/**
 * ObservationForms component
 *
 * A component that displays a search interface for observation forms and a list of selected forms.
 * It allows users to search for observation forms by name, select them, and add them to the consultation.
 *
 * Features:
 * - ComboBox-based search interface
 * - Real-time filtering by form name
 * - Visual distinction for already-added forms
 * - Alphabetical ordering of search results
 * - Internationalization support
 * - Error handling and loading states
 */
const ObservationForms: React.FC<ObservationFormsProps> = React.memo(
  ({
    onFormSelect,
    selectedForms = [],
    onRemoveForm,
    pinnedForms,
    updatePinnedForms,
    isPinnedFormsLoading,
  }) => {
    const { t } = useTranslation();
    const [searchTerm, setSearchTerm] = useState('');

    const { forms: allForms, isLoading: isAllFormsLoading } =
      useObservationFormsSearch();

    const {
      forms: availableForms,
      isLoading: isSearchLoading,
      error: searchError,
    } = useObservationFormsSearch(searchTerm);

    // Validate and filter available forms - handle malformed data
    const validatedAvailableForms = useMemo(() => {
      return allForms.filter((form) => {
        // Check required properties
        if (!form || typeof form !== 'object') return false;
        if (!form.uuid || typeof form.uuid !== 'string') return false;
        if (!form.name || typeof form.name !== 'string') return false;
        return true;
      });
    }, [allForms]);

    // Use API names for filtering (these match the actual form names from backend)
    const defaultPinnedForms = validatedAvailableForms.filter((form) =>
      DEFAULT_FORM_API_NAMES.includes(form.name),
    );

    // Filter orphaned pinned forms - remove forms that are pinned but no longer available
    const validUserPinnedForms = useMemo(() => {
      return pinnedForms.filter((pinnedForm) => {
        return validatedAvailableForms.some(
          (availableForm) => availableForm.uuid === pinnedForm.uuid,
        );
      });
    }, [pinnedForms, validatedAvailableForms]);

    // Merge with user-pinned forms (avoid duplicates)
    const userPinnedUuids = validUserPinnedForms.map((f) => f.uuid);

    // Step 1: Get default forms that user hasn't pinned, sorted alphabetically
    const sortedDefaultForms = defaultPinnedForms
      .filter((f) => !userPinnedUuids.includes(f.uuid))
      .sort((a, b) => a.name.localeCompare(b.name));

    // Step 2: Get user-pinned forms, sorted alphabetically
    const sortedUserPinnedForms = [...validUserPinnedForms].sort((a, b) =>
      a.name.localeCompare(b.name),
    );

    // Step 3: Combine - defaults first, then user-pinned
    const allPinnedForms = [...sortedDefaultForms, ...sortedUserPinnedForms];

    const handleSearch = useCallback((searchQuery: string) => {
      setSearchTerm(searchQuery);
    }, []);

    const handleOnChange = useCallback(
      (data: {
        selectedItem?: { id: string; label: string; disabled?: boolean } | null;
      }) => {
        const selectedItem = data.selectedItem;
        if (!selectedItem?.id || selectedItem.disabled) {
          return;
        }
        const form = availableForms.find(
          (f: ObservationForm) => f.uuid === selectedItem.id,
        );
        if (form) {
          onFormSelect?.(form);
        }
      },
      [availableForms, onFormSelect],
    );

    const searchResults = useMemo(() => {
      if (isSearchLoading) {
        return [
          {
            id: '',
            label: t('OBSERVATION_FORMS_LOADING_FORMS'),
            disabled: true,
          },
        ];
      }

      if (searchError) {
        return [
          {
            id: '',
            label: t('OBSERVATION_FORMS_ERROR_LOADING_FORMS'),
            disabled: true,
          },
        ];
      }

      if (availableForms.length === 0 && searchTerm.length > 0) {
        return [
          {
            id: '',
            label: t('OBSERVATION_FORMS_NO_FORMS_FOUND'),
            disabled: true,
          },
        ];
      }

      if (availableForms.length === 0) {
        return [
          {
            id: '',
            label: 'No forms available',
            disabled: true,
          },
        ];
      }

      // Map forms to ComboBox items with proper labeling for already selected forms
      const results = availableForms.map((form: ObservationForm) => {
        const isAlreadySelected = selectedForms.some(
          (selected: ObservationForm) => selected.uuid === form.uuid,
        );

        return {
          id: form.uuid,
          label: isAlreadySelected
            ? `${form.name} (${t('OBSERVATION_FORMS_FORM_ALREADY_ADDED')})`
            : form.name,
          disabled: isAlreadySelected,
        };
      });

      return results;
    }, [
      isSearchLoading,
      searchError,
      searchTerm,
      availableForms,
      selectedForms,
      t,
    ]);

    return (
      <Tile
        className={styles.observationFormsTile}
        data-testid="observation-forms-tile"
      >
        <div
          className={styles.observationFormsTitle}
          data-testid="observation-forms-title"
        >
          {t('OBSERVATION_FORMS_SECTION_TITLE')}
        </div>

        <div data-testid="observation-forms-search-section">
          <ComboBox
            id="observation-forms-search"
            placeholder={t('OBSERVATION_FORMS_SEARCH_PLACEHOLDER')}
            items={searchResults}
            itemToString={(item) => item?.label ?? ''}
            onChange={handleOnChange}
            onInputChange={handleSearch}
            size="md"
            autoAlign
            disabled={isSearchLoading}
            aria-label={t('OBSERVATION_FORMS_SEARCH_ARIA_LABEL')}
            data-testid="observation-forms-search-combobox"
          />
        </div>

        {selectedForms && selectedForms.length > 0 && (
          <div data-testid="added-forms-section">
            <FormCardContainer
              title={t('OBSERVATION_FORMS_ADDED_FORMS')}
              dataTestId="added-forms-container"
            >
              {selectedForms.map((form: ObservationForm) => (
                <FormCard
                  key={form.uuid}
                  title={form.name}
                  icon="fa-file-lines"
                  actionIcon="fa-times"
                  onOpen={() => onFormSelect?.(form)}
                  onActionClick={() => onRemoveForm?.(form.uuid)}
                  dataTestId={`selected-form-${form.uuid}`}
                  ariaLabel={`Open ${form.name} form`}
                />
              ))}
            </FormCardContainer>
          </div>
        )}

        <div data-testid="pinned-forms-section">
          <FormCardContainer
            title={t('DEFAULT_AND_PINNED_FORMS_TITLE')}
            showNoFormsMessage={
              !isAllFormsLoading &&
              allPinnedForms.length === 0 &&
              defaultPinnedForms.length === 0
            }
            noFormsMessage={t('DEFAULT_AND_PINNED_FORMS_NO_FORMS_FOUND')}
            dataTestId="pinned-forms-container"
          >
            {isAllFormsLoading || isPinnedFormsLoading ? (
              <SkeletonText
                width="100%"
                lineCount={3}
                testId="pinned-forms-skeleton"
              />
            ) : (
              allPinnedForms.map((form: ObservationForm) => (
                <FormCard
                  key={form.uuid}
                  title={form.name}
                  icon="fa-file-lines"
                  actionIcon={
                    !DEFAULT_FORM_API_NAMES.includes(form.name)
                      ? 'fa-thumbtack'
                      : undefined
                  }
                  onOpen={() => onFormSelect?.(form)}
                  onActionClick={() => {
                    const newPinnedForms = pinnedForms.filter(
                      (f) => f.uuid !== form.uuid,
                    );
                    updatePinnedForms(newPinnedForms);
                  }}
                  dataTestId={`pinned-form-${form.uuid}`}
                  ariaLabel={`Open ${form.name} form`}
                />
              ))
            )}
          </FormCardContainer>
        </div>
      </Tile>
    );
  },
);

ObservationForms.displayName = 'ObservationForms';

export default ObservationForms;
