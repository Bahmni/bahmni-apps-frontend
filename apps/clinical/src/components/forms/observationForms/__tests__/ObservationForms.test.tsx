import { ObservationForm } from '@bahmni-frontend/bahmni-services';
import { render, screen, fireEvent } from '@testing-library/react';
import ObservationForms from '../ObservationForms';

// Mock the translation hook
jest.mock('react-i18next', () => ({
  useTranslation: jest.fn(() => ({
    t: jest.fn((key) => `translated_${key}`),
  })),
}));

// Mock the observation forms search hook
jest.mock('../../../../hooks/useObservationFormsSearch');

// Mock Carbon components
jest.mock('@carbon/react', () => ({
  ComboBox: jest.fn(
    ({ items, onChange, onInputChange, disabled, placeholder }) => (
      <div data-testid="combobox">
        <input
          data-testid="combobox-input"
          placeholder={placeholder}
          disabled={disabled}
          onChange={(e) => onInputChange?.(e.target.value)}
        />
        <div data-testid="combobox-items">
          {items?.map(
            (item: { id: string; label: string; disabled?: boolean }) => (
              <button
                key={item.id}
                data-testid={`combobox-item-${item.id}`}
                disabled={item.disabled}
                onClick={() => onChange?.({ selectedItem: item })}
              >
                {item.label}
              </button>
            ),
          )}
        </div>
      </div>
    ),
  ),
  Tile: jest.fn(({ children, className }) => (
    <div data-testid="tile" className={className}>
      {children}
    </div>
  )),
}));

// Mock common components
jest.mock('@bahmni-frontend/bahmni-design-system', () => ({
  ...jest.requireActual('@bahmni-frontend/bahmni-design-system'),
  ComboBox: jest.fn(
    ({ items, onChange, onInputChange, disabled, placeholder }) => (
      <div data-testid="combobox">
        <input
          data-testid="combobox-input"
          placeholder={placeholder}
          disabled={disabled}
          onChange={(e) => onInputChange?.(e.target.value)}
        />
        <div data-testid="combobox-items">
          {items?.map(
            (item: { id: string; label: string; disabled?: boolean }) => (
              <button
                key={item.id}
                data-testid={`combobox-item-${item.id}`}
                disabled={item.disabled}
                onClick={() => onChange?.({ selectedItem: item })}
              >
                {item.label}
              </button>
            ),
          )}
        </div>
      </div>
    ),
  ),
  Tile: jest.fn(({ children, className }) => (
    <div data-testid="tile" className={className}>
      {children}
    </div>
  )),
  BoxWHeader: jest.fn(({ title, children, className }) => (
    <div data-testid="box-with-header" className={className}>
      <div data-testid="box-header">{title}</div>
      <div data-testid="box-content">{children}</div>
    </div>
  )),
  SelectedItem: jest.fn(({ children, onClose, className }) => (
    <div data-testid="selected-item" className={className}>
      {children}
      <button data-testid="selected-item-close" onClick={() => onClose?.()}>
        Ã—
      </button>
    </div>
  )),
  Icon: jest.fn(({ id, name }) => (
    <div data-testid={`bahmni-icon-${id}`} data-icon-name={name}>
      Icon
    </div>
  )),
}));

// SelectedItem is already mocked as part of the design system mock above

// BahmniIcon is already mocked as part of the design system mock above

describe('ObservationForms', () => {
  const mockForms: ObservationForm[] = [
    {
      name: 'Admission Letter',
      uuid: 'form-1',
      id: 1,
      privileges: [],
    },
    {
      name: 'Death Note',
      uuid: 'form-2',
      id: 2,
      privileges: [],
    },
  ];

  const defaultProps = {
    onFormSelect: jest.fn(),
    selectedForms: [],
    onRemoveForm: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();

    // Default mock for useObservationFormsSearch
    const mockUseObservationFormsSearch = jest.requireMock(
      '../../../../hooks/useObservationFormsSearch',
    ).default;
    mockUseObservationFormsSearch.mockReturnValue({
      forms: mockForms,
      isLoading: false,
      error: null,
    });
  });

  describe('Rendering and Structure', () => {
    it('should render the ObservationForms component', () => {
      render(<ObservationForms {...defaultProps} />);

      expect(screen.getByTestId('tile')).toBeInTheDocument();
      expect(
        screen.getByText('translated_OBSERVATION_FORMS_SECTION_TITLE'),
      ).toBeInTheDocument();
      expect(screen.getByTestId('combobox')).toBeInTheDocument();
    });

    it('should render search placeholder correctly', () => {
      render(<ObservationForms {...defaultProps} />);

      const input = screen.getByTestId('combobox-input');
      expect(input).toHaveAttribute(
        'placeholder',
        'translated_OBSERVATION_FORMS_SEARCH_PLACEHOLDER',
      );
    });

    it('should match the snapshot', () => {
      const { container } = render(<ObservationForms {...defaultProps} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Form Search and Selection', () => {
    it('should display available forms in dropdown', () => {
      render(<ObservationForms {...defaultProps} />);

      expect(screen.getByTestId('combobox-item-form-1')).toBeInTheDocument();
      expect(screen.getByTestId('combobox-item-form-2')).toBeInTheDocument();
      expect(screen.getByText('Admission Letter')).toBeInTheDocument();
      expect(screen.getByText('Death Note')).toBeInTheDocument();
    });

    it('should call onFormSelect when a form is selected from dropdown', () => {
      const mockOnFormSelect = jest.fn();
      render(
        <ObservationForms {...defaultProps} onFormSelect={mockOnFormSelect} />,
      );

      const formButton = screen.getByTestId('combobox-item-form-1');
      fireEvent.click(formButton);

      expect(mockOnFormSelect).toHaveBeenCalledWith(mockForms[0]);
    });

    it('should handle search input changes', () => {
      const mockUseObservationFormsSearch = jest.requireMock(
        '../../../../hooks/useObservationFormsSearch',
      ).default;
      render(<ObservationForms {...defaultProps} />);

      const input = screen.getByTestId('combobox-input');
      fireEvent.change(input, { target: { value: 'History' } });

      // Verify the search hook was called with the search term
      expect(mockUseObservationFormsSearch).toHaveBeenCalled();
    });

    it('should not call onFormSelect for disabled items', () => {
      const mockOnFormSelect = jest.fn();
      const selectedForms = [mockForms[0]]; // First form is already selected

      render(
        <ObservationForms
          {...defaultProps}
          onFormSelect={mockOnFormSelect}
          selectedForms={selectedForms}
        />,
      );

      const disabledButton = screen.getByTestId('combobox-item-form-1');
      expect(disabledButton).toBeDisabled();

      fireEvent.click(disabledButton);
      expect(mockOnFormSelect).not.toHaveBeenCalled();
    });
  });

  describe('Selected Forms Display', () => {
    it('should not show Added Forms section when no forms are selected', () => {
      render(<ObservationForms {...defaultProps} selectedForms={[]} />);

      expect(screen.queryByTestId('box-with-header')).not.toBeInTheDocument();
    });

    it('should show Added Forms section when forms are selected', () => {
      const selectedForms = [mockForms[0]];
      render(
        <ObservationForms {...defaultProps} selectedForms={selectedForms} />,
      );

      expect(screen.getByTestId('box-with-header')).toBeInTheDocument();
      expect(screen.getByTestId('box-header')).toHaveTextContent(
        'translated_OBSERVATION_FORMS_ADDED_FORMS',
      );
      expect(screen.getByTestId('selected-item')).toBeInTheDocument();
    });

    it('should display selected form details correctly', () => {
      const selectedForms = [mockForms[0]];
      render(
        <ObservationForms {...defaultProps} selectedForms={selectedForms} />,
      );

      expect(screen.getByText('Admission Letter')).toBeInTheDocument();
      expect(
        screen.getByTestId('bahmni-icon-fa-file-lines'),
      ).toBeInTheDocument();
    });

    it('should call onFormSelect when clicking on selected form', () => {
      const mockOnFormSelect = jest.fn();
      const selectedForms = [mockForms[0]];

      render(
        <ObservationForms
          {...defaultProps}
          onFormSelect={mockOnFormSelect}
          selectedForms={selectedForms}
        />,
      );

      const formContent = screen.getByText('Admission Letter').closest('div');
      fireEvent.click(formContent!);

      expect(mockOnFormSelect).toHaveBeenCalledWith(mockForms[0]);
    });

    it('should call onRemoveForm when clicking close button', () => {
      const mockOnRemoveForm = jest.fn();
      const selectedForms = [mockForms[0]];

      render(
        <ObservationForms
          {...defaultProps}
          onRemoveForm={mockOnRemoveForm}
          selectedForms={selectedForms}
        />,
      );

      const closeButton = screen.getByTestId('selected-item-close');
      fireEvent.click(closeButton);

      expect(mockOnRemoveForm).toHaveBeenCalledWith('form-1');
    });
  });

  describe('Form State Management', () => {
    it('should show already selected forms as disabled in dropdown', () => {
      const selectedForms = [mockForms[0]];
      render(
        <ObservationForms {...defaultProps} selectedForms={selectedForms} />,
      );

      const selectedFormButton = screen.getByTestId('combobox-item-form-1');
      const unselectedFormButton = screen.getByTestId('combobox-item-form-2');

      expect(selectedFormButton).toBeDisabled();
      expect(unselectedFormButton).not.toBeDisabled();

      expect(
        screen.getByText(
          'Admission Letter (translated_OBSERVATION_FORMS_FORM_ALREADY_ADDED)',
        ),
      ).toBeInTheDocument();
      expect(screen.getByText('Death Note')).toBeInTheDocument();
    });

    it('should display multiple selected forms', () => {
      const selectedForms = mockForms;
      render(
        <ObservationForms {...defaultProps} selectedForms={selectedForms} />,
      );

      const selectedItems = screen.getAllByTestId('selected-item');
      expect(selectedItems).toHaveLength(2);

      expect(screen.getByText('Admission Letter')).toBeInTheDocument();
      expect(screen.getByText('Death Note')).toBeInTheDocument();
    });
  });

  describe('Loading and Error States', () => {
    it('should show loading state in dropdown', () => {
      const mockUseObservationFormsSearch = jest.requireMock(
        '../../../../hooks/useObservationFormsSearch',
      ).default;
      mockUseObservationFormsSearch.mockReturnValue({
        forms: [],
        isLoading: true,
        error: null,
      });

      render(<ObservationForms {...defaultProps} />);

      expect(
        screen.getByText('translated_OBSERVATION_FORMS_LOADING_FORMS'),
      ).toBeInTheDocument();
      expect(screen.getByTestId('combobox-input')).toBeDisabled();
    });

    it('should show error state in dropdown', () => {
      const mockUseObservationFormsSearch = jest.requireMock(
        '../../../../hooks/useObservationFormsSearch',
      ).default;
      mockUseObservationFormsSearch.mockReturnValue({
        forms: [],
        isLoading: false,
        error: new Error('Failed to load forms'),
      });

      render(<ObservationForms {...defaultProps} />);

      expect(
        screen.getByText('translated_OBSERVATION_FORMS_ERROR_LOADING_FORMS'),
      ).toBeInTheDocument();
    });

    it('should show no forms found message when search returns empty results', () => {
      const mockUseObservationFormsSearch = jest.requireMock(
        '../../../../hooks/useObservationFormsSearch',
      ).default;
      mockUseObservationFormsSearch.mockReturnValue({
        forms: [],
        isLoading: false,
        error: null,
      });

      render(<ObservationForms {...defaultProps} />);

      // Simulate search input to trigger the "no forms found" state
      const input = screen.getByTestId('combobox-input');
      fireEvent.change(input, { target: { value: 'nonexistent form' } });

      expect(
        screen.getByText('translated_OBSERVATION_FORMS_NO_FORMS_FOUND'),
      ).toBeInTheDocument();
    });

    it('should show no forms available message when no forms exist', () => {
      const mockUseObservationFormsSearch = jest.requireMock(
        '../../../../hooks/useObservationFormsSearch',
      ).default;
      mockUseObservationFormsSearch.mockReturnValue({
        forms: [],
        isLoading: false,
        error: null,
      });

      render(<ObservationForms {...defaultProps} />);

      expect(screen.getByText('No forms available')).toBeInTheDocument();
    });
  });

  describe('Edge Cases and Branch Coverage', () => {
    it('should not call onFormSelect when selectedItem is null', () => {
      const mockOnFormSelect = jest.fn();
      render(
        <ObservationForms {...defaultProps} onFormSelect={mockOnFormSelect} />,
      );

      // Simulate ComboBox onChange with null selectedItem
      const ComboBox = jest.requireMock('@carbon/react').ComboBox;
      const lastCall = ComboBox.mock.calls[ComboBox.mock.calls.length - 1];
      const onChange = lastCall[0].onChange;

      onChange({ selectedItem: null });

      expect(mockOnFormSelect).not.toHaveBeenCalled();
    });

    it('should not call onFormSelect when selectedItem has no id', () => {
      const mockOnFormSelect = jest.fn();
      render(
        <ObservationForms {...defaultProps} onFormSelect={mockOnFormSelect} />,
      );

      // Simulate ComboBox onChange with selectedItem without id
      const ComboBox = jest.requireMock('@carbon/react').ComboBox;
      const lastCall = ComboBox.mock.calls[ComboBox.mock.calls.length - 1];
      const onChange = lastCall[0].onChange;

      onChange({
        selectedItem: { id: '', label: 'Loading...', disabled: true },
      });

      expect(mockOnFormSelect).not.toHaveBeenCalled();
    });

    it('should handle itemToString function correctly', () => {
      render(<ObservationForms {...defaultProps} />);

      // Get the itemToString function from ComboBox mock
      const ComboBox = jest.requireMock('@carbon/react').ComboBox;
      const lastCall = ComboBox.mock.calls[ComboBox.mock.calls.length - 1];
      const itemToString = lastCall[0].itemToString;

      // Test with valid item
      expect(itemToString({ label: 'Test Form' })).toBe('Test Form');

      // Test with null item (covers line 142 fallback)
      expect(itemToString(null)).toBe('');

      // Test with undefined item
      expect(itemToString(undefined)).toBe('');

      // Test with item without label
      expect(itemToString({})).toBe('');
    });

    it('should handle optional callbacks gracefully', () => {
      render(<ObservationForms />);

      // Should not throw when callbacks are not provided
      const formButton = screen.getByTestId('combobox-item-form-1');
      expect(() => fireEvent.click(formButton)).not.toThrow();
    });

    it('should handle onFormSelect being undefined', () => {
      // Test the branch where onFormSelect is undefined (line 59)
      render(<ObservationForms onFormSelect={undefined} />);

      const formButton = screen.getByTestId('combobox-item-form-1');
      expect(() => fireEvent.click(formButton)).not.toThrow();
    });
  });

  describe('Search Functionality Edge Cases', () => {
    it('should handle empty search term correctly', () => {
      render(<ObservationForms {...defaultProps} />);

      const input = screen.getByTestId('combobox-input');
      fireEvent.change(input, { target: { value: '' } });

      // Should show all available forms when search is empty
      expect(screen.getByTestId('combobox-item-form-1')).toBeInTheDocument();
      expect(screen.getByTestId('combobox-item-form-2')).toBeInTheDocument();
    });

    it('should handle whitespace-only search terms', () => {
      render(<ObservationForms {...defaultProps} />);

      const input = screen.getByTestId('combobox-input');
      fireEvent.change(input, { target: { value: '   ' } });

      // Should treat whitespace-only as empty search
      expect(screen.getByTestId('combobox-item-form-1')).toBeInTheDocument();
    });

    it('should handle special characters in search', () => {
      render(<ObservationForms {...defaultProps} />);

      const input = screen.getByTestId('combobox-input');
      fireEvent.change(input, { target: { value: '@#$%' } });

      // Should not crash with special characters
      expect(input).toHaveValue('@#$%');
    });

    it('should handle very long search terms', () => {
      render(<ObservationForms {...defaultProps} />);

      const longSearchTerm = 'a'.repeat(1000);
      const input = screen.getByTestId('combobox-input');
      fireEvent.change(input, { target: { value: longSearchTerm } });

      expect(input).toHaveValue(longSearchTerm);
    });
  });
  describe('Internationalization Support', () => {
    it('should use translation keys for all user-facing text', () => {
      render(<ObservationForms {...defaultProps} />);

      // Check that translation function is called with correct keys
      expect(screen.getByText('translated_OBSERVATION_FORMS_SECTION_TITLE')).toBeInTheDocument();
      expect(screen.getByDisplayValue('')).toHaveAttribute(
        'placeholder',
        'translated_OBSERVATION_FORMS_SEARCH_PLACEHOLDER'
      );
    });
  });
  describe('Form Selection Edge Cases', () => {
    it('should handle case where form is not found in availableForms', () => {
      const mockOnFormSelect = jest.fn();
      render(
        <ObservationForms {...defaultProps} onFormSelect={mockOnFormSelect} />,
      );

      // Simulate ComboBox onChange with a selectedItem that doesn't exist in availableForms
      const ComboBox = jest.requireMock('@carbon/react').ComboBox;
      const lastCall = ComboBox.mock.calls[ComboBox.mock.calls.length - 1];
      const onChange = lastCall[0].onChange;

      // Use an ID that doesn't exist in mockForms
      onChange({
        selectedItem: { id: 'non-existent-form-id', label: 'Non-existent Form', disabled: false },
      });

      // Should not call onFormSelect when form is not found (covers line 62 branch)
      expect(mockOnFormSelect).not.toHaveBeenCalled();
    });

    it('should handle selectedItem with disabled true', () => {
      const mockOnFormSelect = jest.fn();
      render(
        <ObservationForms {...defaultProps} onFormSelect={mockOnFormSelect} />,
      );

      const ComboBox = jest.requireMock('@carbon/react').ComboBox;
      const lastCall = ComboBox.mock.calls[ComboBox.mock.calls.length - 1];
      const onChange = lastCall[0].onChange;

      // Test with disabled selectedItem
      onChange({
        selectedItem: { id: 'form-1', label: 'Test Form', disabled: true },
      });

      expect(mockOnFormSelect).not.toHaveBeenCalled();
    });
  });
});
