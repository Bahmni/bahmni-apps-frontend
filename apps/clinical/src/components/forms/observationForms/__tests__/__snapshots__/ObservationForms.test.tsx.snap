// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservationForms Rendering and Structure should match the snapshot 1`] = `
<div>
  <div
    class="observationFormsTile"
    data-testid="tile"
  >
    <div
      class="observationFormsTitle"
    >
      translated_OBSERVATION_FORMS_SECTION_TITLE
    </div>
    <div
      data-testid="combobox"
    >
      <input
        data-testid="combobox-input"
        placeholder="translated_OBSERVATION_FORMS_SEARCH_PLACEHOLDER"
      />
      <div
        data-testid="combobox-items"
      >
        <button
          data-testid="combobox-item-form-1"
        >
          Admission Letter
        </button>
        <button
          data-testid="combobox-item-form-2"
        >
          Death Note
        </button>
      </div>
    </div>
  </div>
</div>
`;
